### PR TITLE
[FLINK-30528] Reconcile other changes if upgrade is not available

### PR DESCRIPTION
**Pull request by gyfora**
## What is the purpose of the change

Trigger other reconciliation actions even if upgrademode is currently not available to not cause a stuck upgrade loop.

## Brief change log

 - Improved spec reconciliation logic
 - Unit tests

## Verifying this change

Added new reconciler unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java [hunk 1]
```diff
@@ -706,4 +707,47 @@ public void testRestartUnhealthyEvent() throws Exception {
         reconciler.reconcile(deployment, context);
         Assertions.assertEquals(MSG_RESTART_UNHEALTHY, eventCollector.events.remove().getMessage());
     }
+
+    @Test
+    public void testReconcileIfUpgradeModeNotAvailable() throws Exception {
```
Comment (morhidi) [inline flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java:712]: Could you please add comments to the test @gyfora? It's really hard to follow what it actually does.

Comment (gyfora) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java:712]: done

**Comment by morhidi** [comment]
Thanks @gyfora. +1 for merge
